File: src/csharp/Foundgine.Runtime/Api/IFoundgineMutations.cs
using Foundgine.Core.Execution;
using Foundgine.Core.Execution.Mutation;
using Foundgine.Core.Semantic.Mutation;
using ExecutionContext = Foundgine.Core.Execution.ExecutionContext;
using Foundgine.Core.Semantic.Security.Execution;

namespace Foundgine.Runtime;

public interface IFoundgineMutations
{
    MutationDryRunResult DryRun(SemanticMutationRequest request);

    Task<MutationExecutionResult> ExecuteAsync(
        SemanticMutationRequest request,
        ExecutionContext? context = null,
        CancellationToken cancellationToken = default);

    MutationPlanApproval Approve(SemanticMutationRequest request, string approvedBy);

    Task<MutationExecutionResult> ExecuteApprovedAsync(
        MutationPlanApproval approval,
        ExecutionContext? context = null,
        CancellationToken cancellationToken = default);
}

public sealed record SemanticMutationRequest(
    SemanticMutationOperationGraph Graph,
    SecurityExecutionContext? Security = null);

public sealed record MutationDryRunResult(
    string PlanFingerprint,
    IReadOnlyList<MutationPlanOperation> Operations,
    IReadOnlyList<string> Effects);

public sealed record MutationPlanOperation(
    int Index,
    string Entity,
    string Kind,
    IReadOnlyList<string> Fields,
    IReadOnlyList<string> ReturnFields);

public sealed record MutationPlanApproval(
    SemanticMutationRequest Request,
    string ApprovalId,
    string PlanFingerprint,
    string ApprovedBy,
    DateTimeOffset ApprovedAt);

public sealed record MutationExecutionResult(
    MutationBatchResult Result,
    string PlanFingerprint,
    string ResultFingerprint,
    string? ApprovalId,
    string? ApprovedBy);

File: src/java/foundgine-runtime/src/main/java/com/foundgine/runtime/api/IFoundgineMutations.java
package com.foundgine.runtime.api;

import com.foundgine.core.execution.ExecutionContext;
import com.foundgine.core.execution.mutation.MutationBatchResult;
import com.foundgine.core.semantic.mutation.*;
import com.foundgine.core.semantic.security.execution.SecurityExecutionContext;
import java.time.OffsetDateTime;
import java.util.*;
import java.util.concurrent.CompletableFuture;

public interface IFoundgineMutations {
	MutationDryRunResult dryRun(SemanticMutationRequest request);

	CompletableFuture<MutationExecutionResult> executeAsync(SemanticMutationRequest request, ExecutionContext context);

	MutationPlanApproval approve(SemanticMutationRequest request, String approvedBy);

	CompletableFuture<MutationExecutionResult> executeApprovedAsync(MutationPlanApproval approval,
			ExecutionContext context);

	record SemanticMutationRequest(SemanticMutationOperationGraph graph, SecurityExecutionContext security) {
		public SemanticMutationRequest(SemanticMutationOperationGraph graph) {
			this(graph, null);
		}
	}

	record MutationDryRunResult(String planFingerprint, List<MutationPlanOperation> operations, List<String> effects) {
	}

	record MutationPlanOperation(int index, String entity, String kind, List<String> fields,
			List<String> returnFields) {
	}

	record MutationPlanApproval(SemanticMutationRequest request, String approvalId, String planFingerprint,
			String approvedBy, OffsetDateTime approvedAt) {
	}

	record MutationExecutionResult(MutationBatchResult result, String planFingerprint, String resultFingerprint,
			String approvalId, String approvedBy) {
	}
}
